speaker/bgp_controller.go

```diff
@@ -80,6 +81,44 @@ newPeers:
 	return c.syncPeers(l)
 }
 
+// nodeHasHealthyEndpoint return true if this node has at least one healthy endpoint.
+func nodeHasHealthyEndpoint(eps *v1.Endpoints, node string) bool {
+	ready := map[string]bool{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			if ep.NodeName == nil || *ep.NodeName != node {
+				continue
+			}
+			if _, ok := ready[ep.IP]; !ok {
+				// Only set true if nothing else has expressed an
+				// opinion. This means that false will take precedence
+				// if there's any unready ports for a given endpoint.
+				ready[ep.IP] = true
+			}
+		}
+		for _, ep := range subset.NotReadyAddresses {
+			ready[ep.IP] = false
+		}
+	}
+
+	for _, r := range ready {
+		if r {
+			// At least one fully healthy endpoint on this machine.
+			return true
+		}
+	}
+	return false
+}
+
+func (c *bgpController) ShouldAnnounce(l log.Logger, name string, svc *v1.Service, eps *v1.Endpoints) string {
```

Comment:
Nit: I tend to prefer an integer enum for this kind of thing rather than a "magic" string, but up to you.

Reply:
Hrm. Tried this, but it just ended up making the rest of the code less clear. I agree that I don't like magic strings though, so that suggests this interface needs fixing.